PebbleSharp.Core/AppMessage/AppMessagePacket.cs

```diff
@@ -0,0 +1,524 @@
+
+					IAppMessageDictionaryEntry entry = null;
+					if (t == (byte)PackedType.Bytes)
+					{
+						entry = new AppMessageBytes() { Value = bytes.Skip(index).Take(l).ToArray() };
+					}
+					else if (t == (byte)PackedType.Signed)
+					{
+						if (l == 1)
+						{
+							entry = new AppMessageInt8() { Value = Convert.ToSByte(bytes[index]) };
+						}
+						else if (l == 2)
+						{
+							entry = new AppMessageInt16() { Value = BitConverter.ToInt16(bytes, index) };
+						}
+						else if (l == 4)
+						{
+							entry = new AppMessageInt32() { Value = BitConverter.ToInt32(bytes, index) };
+						}
+						else
+						{
+							throw new InvalidOperationException("Invalid signed integer length");
+						}
+					}
+					else if (t == (byte)PackedType.String)
+					{
+						entry = new AppMessageString() { Value = System.Text.Encoding.UTF8.GetString(bytes, index, l) };
+					}
+					else if (t == (byte)PackedType.Unsigned)
+					{
+						if (l == 1)
+						{
+							entry = new AppMessageUInt8() { Value = bytes[index] };
+						}
+						else if (l == 2)
+						{
+							entry = new AppMessageUInt16() { Value = BitConverter.ToUInt16(bytes, index) };
+						}
+						else if (l == 4)
+						{
+							entry = new AppMessageUInt32() { Value = BitConverter.ToUInt32(bytes, index) };
+						}
+						else
+						{
+							throw new InvalidOperationException("Invalid signed integer length");
+						}
+					}
+					else
+					{
+						throw new InvalidOperationException("Unknown tuple type");
+					}
+					index += l;
+					entry.Key = k;
+					Values.Add(entry);
+				}
+			}
+		}
+
+        public IList<IAppMessageDictionaryEntry> Values { get; set; }
+
+        public byte[] GetBytes()
+        {
+            if (Values != null && Values.Any())
+            {
+                var bytes = new List<byte>();
+                bytes.Add(Command);
+                bytes.Add(TransactionId);
+                bytes.AddRange(ApplicationId.Data);
+                bytes.Add((byte)Values.Count);
+                foreach (var tuple in Values)
+                {
+                    bytes.AddRange(tuple.PackedBytes);
+                }
+                return bytes.ToArray();
+            }
+            else
+            {
+                return new byte[0];
+            }
+        }
+    }
+
+    public interface IAppMessageDictionaryEntry
+    {
+        uint Key { get; set; }
+        PackedType PackedType { get; }
+        ushort Length { get; }
+        byte[] ValueBytes { get; set; }
+        byte[] PackedBytes { get; }
+    }
+
+    public enum Command:byte
+    {
+        Push=1
+    }
+
+    public enum PackedType:byte
+    {
+        Bytes=0,
+        String =1,
+        Unsigned =2,
+        Signed = 3
+    }
+
+    public abstract class AppMessageDictionaryEntry<T> : IAppMessageDictionaryEntry
+    {
+        public uint Key { get; set; }
+        public abstract PackedType PackedType { get; }
+        public abstract ushort Length { get; }
```

Comment:
sizeof doesn't work with type params (TIL, I thought it would work too), however marshal.sizeof does.  I was torn between cleaning up the code vs using reflection at run time and the associated perf drain, I went with the cleaner code and marshal.sizeof.